src/kernels/common/baseJupyterSession.ts

```diff
@@ -175,44 +174,33 @@ export abstract class BaseJupyterSession implements IJupyterSession {
             return;
         }
 
-        // Start the restart session now in case it wasn't started
-        if (!this.restartSessionPromise) {
-            this.startRestartSession(false);
-        }
+        traceInfo(`Restarting ${this.session?.kernel?.id}`);
 
-        // Just kill the current session and switch to the other
-        if (this.restartSessionPromise) {
-            traceInfo(`Restarting ${this.session?.kernel?.id}`);
+        // Save old state for shutdown
+        const oldSession = this.session;
+        const oldStatusHandler = this.statusHandler;
 
-            // Save old state for shutdown
-            const oldSession = this.session;
-            const oldStatusHandler = this.statusHandler;
+        // TODO? Why aren't we killing this old session here now?
+        // We should, If we're restarting and it fails, how is it ok to
+        // keep the old session (user could be restarting for a number of reasons).
 
-            // TODO? Why aren't we killing this old session here now?
-            // We should, If we're restarting and it fails, how is it ok to
-            // keep the old session (user could be restarting for a number of reasons).
+        // Just switch to the other session. It should already be ready
 
-            // Just switch to the other session. It should already be ready
-            const newSession = await this.restartSessionPromise.promise;
-            this.setSession(newSession);
+        // Start the restart session now in case it wasn't started
+        const newSession = await this.startRestartSession(false);
+        this.setSession(newSession);
 
-            if (newSession.kernel) {
-                this.restartSessionUsed(newSession.kernel);
-                traceInfo(`Got new session ${newSession.kernel.id}`);
+        if (newSession.kernel) {
+            traceInfo(`Got new session ${newSession.kernel.id}`);
 
-                // Rewire our status changed event.
-                newSession.statusChanged.connect(this.statusHandler);
-            }
-            this.restartSessionPromise.token.dispose();
-            this.restartSessionPromise = undefined;
-            traceInfo('Started new restart session');
-            if (oldStatusHandler && oldSession) {
-                oldSession.statusChanged.disconnect(oldStatusHandler);
-            }
-            this.shutdownSession(oldSession, undefined, false).ignoreErrors();
-        } else {
-            throw new SessionDisposedError();
```

Comment:
Its impossiblel for this to happen, but because of the types being nullable, we write such code.
Basiscally we initialize a class and then call a method to set it in the right state, a factory class will get rid of this.
We already have a factory class, hopefully it will be easy to move some of the stuff into there.